dio/lib/src/adapters/browser_adapter.dart

```diff
@@ -71,7 +71,9 @@ class BrowserHttpClientAdapter implements HttpClientAdapter {
           xhr.status!,
           headers: xhr.responseHeaders.map((k, v) => MapEntry(k, v.split(','))),
           statusMessage: xhr.statusText,
-          isRedirect: xhr.status == 302 || xhr.status == 301,
+          isRedirect: xhr.status == 302 ||
+              xhr.status == 301 ||
+              options.uri.toString() != xhr.responseUrl,
```

Comment:
Judging from my tests, the XHR will never return 301/302 here as the browser or JS-engine does always follow redirects.